kernel/src/table_properties.rs

```diff
@@ -0,0 +1,373 @@
+//! Delta Table properties. Note this module implements per-table configuration which governs how
+//! table-level capabilities/properties are configured (turned on/off etc.). This is orthogonal to
+//! protocol-level 'table features' which enable or disable reader/writer features (which then
+//! usually must be enabled/configured by table properties).
+//!
+//! For example (from delta's protocol.md): A feature being supported does not imply that it is
+//! active. For example, a table may have the `appendOnly` feature listed in writerFeatures, but it
+//! does not have a table property delta.appendOnly that is set to `true`. In such a case the table
+//! is not append-only, and writers are allowed to change, remove, and rearrange data. However,
+//! writers must know that the table property delta.appendOnly should be checked before writing the
+//! table.
+
+use std::collections::HashMap;
+use std::time::Duration;
+
+use serde::Deserialize;
+
+use crate::expressions::ColumnName;
+use crate::table_features::ColumnMappingMode;
+use crate::DeltaResult;
+
+mod deserialize;
+use deserialize::*;
+
+/// Default num index cols
+pub const DEFAULT_NUM_INDEX_COLS: i32 = 32;
+
+/// Delta table properties. These are parsed from the 'configuration' map in the most recent
+/// 'Metadata' action of a table.
+#[derive(Deserialize, Debug, Clone, Eq, PartialEq, Default)]
+#[serde(default)]
+pub struct TableProperties {
+    /// true for this Delta table to be append-only. If append-only,
+    /// existing records cannot be deleted, and existing values cannot be updated.
+    #[serde(rename = "delta.appendOnly")]
+    #[serde(deserialize_with = "deserialize_bool")]
+    pub append_only: Option<bool>,
+
+    /// true for Delta Lake to automatically optimize the layout of the files for this Delta table.
+    #[serde(rename = "delta.autoOptimize.autoCompact")]
+    #[serde(deserialize_with = "deserialize_bool")]
+    pub auto_compact: Option<bool>,
+
+    /// true for Delta Lake to automatically optimize the layout of the files for this Delta table
+    /// during writes.
+    #[serde(rename = "delta.autoOptimize.optimizeWrite")]
+    #[serde(deserialize_with = "deserialize_bool")]
+    pub optimize_write: Option<bool>,
+
+    /// Interval (expressed as number of commits) after which a new checkpoint should be created.
+    /// E.g. if checkpoint interval = 10, then a checkpoint should be written every 10 commits.
+    #[serde(rename = "delta.checkpointInterval")]
+    #[serde(deserialize_with = "deserialize_pos_int")]
```

Comment:
nit: would this be easier to read?
```suggestion
    #[serde(deserialize_with = "deserialize_uint")]
```

Reply:
Ah never mind -- it really does enforce positive (`x > 0`), not merely non-negative